File: tests/workers/test_echo.py
# Copyright 2021 Xilinx, Inc.
# Copyright 2022 Advanced Micro Devices, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
import pytest
from helper import run_benchmark
from proteus.predict_api import (
    InferenceRequest,
    InferenceRequestInput,
    InferenceRequestOutput,
)

import proteus


@pytest.mark.usefixtures("load")
class TestEcho:
    """
    Test the Echo worker
    """

    model = "echo"
    parameters = None

    inputs = [3]
    golden_outputs = [4]

    @classmethod
    def construct_request(
        cls, add_id, add_input_parameters, add_request_parameters, multiplier=1
    ):
        """
        Construct a standard request subject to some options

        Args:
            add_id (bool): Add an ID to the request
            add_input_parameters (bool): Add parameters to the input tensors
            add_request_parameters (bool): Add parameters to the request

        Returns:
            dict: The constructed request
        """

        input_0 = InferenceRequestInput()
        input_0.name = "echo"
        input_0.datatype = proteus.DataType.UINT32
        input_0.shape = [1]
        input_0.setUint32Data(np.array([cls.inputs[0]], np.uint32))
        if add_input_parameters:
            parameters = proteus.RequestParameters()
            parameters.put("key", "value")
            input_0.parameters = parameters

        request = InferenceRequest()
        for _ in range(multiplier):
            request.addInputTensor(input_0)

        if add_id:
            request.id = "hello_world"

        if add_request_parameters:
            parameters = proteus.RequestParameters()
            parameters.put("key3", True)
            parameters.put("key4", 1.2)
            request.parameters = parameters

        return request

    @staticmethod
    def add_outputs(request: InferenceRequest, add_parameters):
        """
        Add output tensors to the request. This is optional.

        Args:
            request (dict): The request to add outputs to
            add_parameters (bool): Add parameters to the output tensors

        Returns:
            dict: Updated request
        """
        output_0 = InferenceRequestOutput()
        output_0.name = "echo"
        if add_parameters:
            parameters = proteus.RequestParameters()
            parameters.put("key", "value2")
            output_0.parameters = parameters

        request.addOutputTensor(output_0)
        request.addOutputTensor(output_0)
        return request

    def send_request(self, request, id_present, input_tensors=1):
        """
        Sends the given request to the server and asserts common checks

        Args:
            request (InferenceRequest): Request to send to the server
            input_tensors (int): Number of times the two input tensors are repeated

        Returns:
            dict: Response as a dictionary
        """

        try:
            response = self.rest_client.modelInfer(self.endpoint, request)
        except ConnectionError:
            pytest.fail(
                "Connection to the proteus server ended without response!", False
            )

        assert not response.isError(), response.getError()
        assert response.model == "echo"

        outputs = response.getOutputs()
        assert len(outputs) == 1 * input_tensors

        for i in range(input_tensors):
            output = outputs[i]
            data = output.getUint32Data()
            assert len(data) == 1
            assert data[0] == self.golden_outputs[0]
            assert output.datatype == proteus.DataType.UINT32
            assert output.name == "echo"
            assert output.parameters.empty()
            assert output.shape == [1]

        if id_present:
            assert response.id == "hello_world"

        return response

    def test_echo_0(self):
        """
        Send a request to echo with all optional parameters
        """
        add_id = True
        request = self.construct_request(add_id, True, True)
        request = self.add_outputs(request, True)

        self.send_request(request, add_id)

    def test_echo_1(self):
        """
        Send a request to echo with most optional parameters (remove request
        parameters)
        """
        add_id = True
        request = self.construct_request(add_id, True, False)
        request = self.add_outputs(request, True)

        self.send_request(request, add_id)

    def test_echo_2(self):
        """
        Send a request to echo with some optional parameters (remove all
        parameters)
        """
        add_id = True
        request = self.construct_request(add_id, False, False)
        request = self.add_outputs(request, False)

        self.send_request(request, add_id)

    def test_echo_3(self):
        """
        Send a request to echo with some optional parameters (remove all
        parameters and remove the ID)
        """
        add_id = False
        request = self.construct_request(add_id, False, False)
        request = self.add_outputs(request, False)

        self.send_request(request, add_id)

    def test_echo_4(self):
        """
        Send a request to echo with no optional parameters (remove all
        parameters, remove the ID and output tensors)
        """
        add_id = False
        request = self.construct_request(add_id, False, False)

        self.send_request(request, add_id)

    @pytest.mark.benchmark(group="echo")
    def test_benchmark_echo_0(self, benchmark):
        request = self.construct_request(False, False, False)
        options = {
            "model": self.model,
            "parameters": self.parameters,
            "type": "rest (pytest)",
            "config": "N/A",
        }
        run_benchmark(
            benchmark, "echo_4", self.rest_client.modelInfer, request, **options
        )


# TODO(varunsh): these tests cannot run now with the new Pybind11 API as it
# doesn't allow creating these kinds of bad requests like this. We'd need to
# reimplement them another way to test these cases
# @pytest.mark.usefixtures("load")
# class TestEchoErrors:
#     """
#     Test the Echo worker with error-causing requests
#     """

#     def send_request(self, request, error):
#         """
#         Send the request to echo and check the error message

#         Args:
#             request (dict): The request to send
#             error (str): Error message to expect
#         """
#         try:
#             response = self.rest_client.modelInfer("echo", request)
#         except ConnectionError:
#             pytest.fail(
#                 "Connection to the proteus server ended without response!", False
#             )

#         assert response.error
#         assert response.status_code == 400
#         assert response.error_msg == error

#     def test_no_inputs(self):
#         """
#         Providing no input tensors is an error
#         """
#         request = {"foo": "bar"}

#         self.send_request(request, "No 'inputs' key present in request")

#     def test_bad_inputs_0(self):
#         """
#         Providing non-array input tensors is an error
#         """
#         request = {"inputs": "bar"}

#         self.send_request(request, "'inputs' is not an array")

#     def test_bad_inputs_1(self):
#         """
#         If inputs is an array, each element must be an object (e.g. a dict)
#         """
#         request = {"inputs": [0]}

#         self.send_request(request, "At least one element in 'inputs' is not an obj")

#     def test_bad_inputs_2(self):
#         """
#         Each input tensor must be named
#         """
#         request = {"inputs": [{"foo": "bar"}], "outputs": []}

#         self.send_request(request, "No 'name' key present in request input")

#     def test_bad_inputs_3(self):
#         """
#         Each input tensor must define its shape
#         """
#         request = {"inputs": [{"name": "bar"}], "outputs": []}

#         self.send_request(request, "No 'shape' key present in request input")

#     def test_bad_inputs_4(self):
#         """
#         An input tensor's shape must be specified with a value that can be
#         coerced into a unint64 value
#         """
#         request = {"inputs": [{"name": "bar", "shape": ["foo"]}], "outputs": []}

#         self.send_request(request, "'shape' must be specified by uint64 elements")

#     def test_bad_inputs_5(self):
#         """
#         All input tensors must define their datatype
#         """
#         request = {"inputs": [{"name": "bar", "shape": [1]}], "outputs": []}

#         self.send_request(request, "No 'datatype' key present in request input")

#     def test_bad_inputs_6(self):
#         """
#         Input tensors' datatypes must be one of the supported values
#         """
#         request = {
#             "inputs": [{"name": "bar", "shape": [1], "datatype": "BAD_TYPE"}],
#             "outputs": [],
#         }

#         self.send_request(request, "Unknown datatype: BAD_TYPE")

#     def test_bad_inputs_7(self):
#         """
#         Input tensors must define their input data
#         """
#         request = {
#             "inputs": [{"name": "bar", "shape": [1], "datatype": "UINT32"}],
#             "outputs": [],
#         }

#         self.send_request(request, "No 'data' key present in request input")

File: tests/workers/test_echo.cpp
// Copyright 2022 Xilinx, Inc.
// Copyright 2022 Advanced Micro Devices, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <array>    // for array
#include <chrono>   // for seconds
#include <cstdint>  // for uint32_t
#include <memory>   // for allocator, make_shared, __shared_ptr_...
#include <string>   // for string
#include <thread>   // for sleep_for
#include <vector>   // for vector

#include "amdinfer/amdinfer.hpp"  // for ParameterMap, InferenceRequestInput
#include "gtest/gtest.h"  // for ParamIteratorInterface, Message, Test...

namespace amdinfer {

struct Params {
  bool add_id;
  bool add_input_parameters;
  bool add_request_parameters;
  bool add_outputs;
  int multiplier;
};

class EchoParamFixture : public testing::TestWithParam<Params> {
 public:
  const std::array<int, 1> inputs{3};
  const std::array<int, 1> golden_outputs{4};

  [[nodiscard]] InferenceRequest constructRequest() const {
    const auto params = GetParam();

    InferenceRequestInput input_0;
    input_0.setName("echo");
    input_0.setDatatype(DataType::Uint32);
    input_0.setShape({1});
    // NOLINTNEXTLINE(google-readability-casting)
    input_0.setData((void*)(&(inputs[0])));
    if (params.add_input_parameters) {
      ParameterMap parameters;
      parameters.put("key_0", "value");
      input_0.setParameters(parameters);
    }

    InferenceRequest request;
    for (auto i = 0; i < params.multiplier; i++) {
      request.addInputTensor(input_0);
    }

    if (params.add_outputs) {
      InferenceRequestOutput output;
      output.setName("echo");
      if (params.add_input_parameters) {
        ParameterMap parameters;
        parameters.put("key", "another_value");
        output.setParameters(parameters);
      }
      for (auto i = 0; i < params.multiplier; i++) {
        request.addOutputTensor(output);
      }
    }

    if (params.add_id) {
      request.setID("hello_world");
    }

    if (params.add_request_parameters) {
      const auto key_3 = 1.2;  // arbitrary value
      ParameterMap parameters;
      parameters.put("key_2", true);
      parameters.put("key_3", key_3);
      request.setParameters(parameters);
    }

    return request;
  }

  void validate(InferenceResponse response) {
    const auto params = GetParam();
    auto add_id = params.add_id;
    auto multiplier = params.multiplier;

    ASSERT_FALSE(response.isError());
    EXPECT_EQ(response.getModel(), "echo");
    auto outputs = response.getOutputs();
    EXPECT_EQ(outputs.size(), multiplier);

    for (auto i = 0; i < multiplier; i++) {
      const auto& output = outputs[i];
      const auto* data = static_cast<uint32_t*>(output.getData());
      EXPECT_EQ(data[0], golden_outputs[0]);
      EXPECT_EQ(output.getDatatype(), DataType::Uint32);
      EXPECT_EQ(output.getName(), "echo");
      EXPECT_TRUE(output.getParameters().empty());
      auto shape = output.getShape();
      EXPECT_EQ(shape.size(), 1);
      EXPECT_EQ(shape[0], 1);
    }

    if (add_id) {
      EXPECT_EQ(response.getID(), "hello_world");
    }
  }

  amdinfer::Server server;
};

TEST_P(EchoParamFixture, EchoNative) {  // NOLINT
  amdinfer::NativeClient client(&server);
  const auto endpoint = client.workerLoad("echo", {});

  auto request = this->constructRequest();

  auto response = client.modelInfer(endpoint, request);

  validate(response);

  client.modelUnload(endpoint);
}

#ifdef AMDINFER_ENABLE_GRPC
TEST_P(EchoParamFixture, EchoGrpc) {  // NOLINT
  server.startGrpc(kDefaultGrpcPort);
  auto client =
    amdinfer::GrpcClient("localhost:" + std::to_string(kDefaultGrpcPort));
  while (!client.serverLive()) {
    std::this_thread::sleep_for(std::chrono::seconds(1));
  }

  const auto endpoint = client.workerLoad("echo", {});

  auto request = this->constructRequest();

  auto response = client.modelInfer(endpoint, request);
  validate(response);

  client.modelUnload(endpoint);
}
#endif

// add_id, add_input_parameters, add_request_parameters, add_outputs, multiplier
const std::array kConfigs{
  Params{true, true, true, true, 1},     Params{true, true, false, true, 1},
  Params{true, false, false, true, 1},   Params{false, false, false, true, 1},
  Params{false, false, false, false, 1},
};

// NOLINTNEXTLINE(cert-err58-cpp)
INSTANTIATE_TEST_SUITE_P(Echo, EchoParamFixture, testing::ValuesIn(kConfigs));

}  // namespace amdinfer
